binary-oracle-pair/program/src/instruction.rs

```diff
@@ -0,0 +1,213 @@
+//! Instruction types
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use solana_program::{
+    clock::Slot,
+    instruction::{AccountMeta, Instruction},
+    program_error::ProgramError,
+    pubkey::Pubkey,
+    sysvar,
+};
+
+/// Initialize arguments for pool
+#[repr(C)]
+#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug, Clone)]
+pub struct InitArgs {
+    /// mint end slot
+    pub mint_end_slot: Slot,
+    /// decide end slot
+    pub decide_end_slot: Slot,
+    /// authority nonce
+    pub bump_seed: u8,
+}
+
+/// Instruction definition
+#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug, Clone)]
+pub enum PoolInstruction {
+    /// Initializes a new binary oracle pair pool.
+    ///
+    ///   0. `[w]` Pool account.
+    ///   1. `[]` Authority
+    ///   2. `[]` Decider authority
+    ///   3. `[]` Deposit currency SPL Token mint. Must be initialized.
+    ///   4. `[w]` Deposit token account. Should not be initialized
+    ///   5. `[w]` Token Pass mint. Should not be initialized
+    ///   6. `[w]` Token Fail mint. Should not be initialized
+    ///   7. `[]` Rent sysvar
+    ///   8. `[]` Token program id
+    InitPool(InitArgs),
+
+    ///   Deposit in the pool.
+    ///
+    ///   0. `[]` Pool
+    ///   1. `[]` Authority
+    ///   2. `[w]` Token SOURCE Account, amount is transferable by pool authority with allowances.
```

Comment:
How will it work then ?
To transfer tokens from account transaction should be signed with this authority
And when we call transfer instruction inside the program we can sign it only with program account
Or if we sign `Deposit` instruction with authority key we can call just `invoke()` inside the program to transfer tokens ?

Reply:
> Or if we sign Deposit instruction with authority key we can call just invoke() inside the program to transfer tokens ?

Exactly.  In this particular usecase, it should be ok, since the UI will approve exactly the amount before calling deposit or withdraw.  Either way, the best practice is to approve to an ephemeral keypair, and then sign with that.